src/ngraph/pass/constant_folding.cpp

```diff
@@ -763,3 +766,176 @@ void pass::ConstantFolding::construct_constant_quantize()
     this->add_matcher(
         quantize_matcher, constant_quantize_callback, PassProperty::REQUIRE_STATIC_SHAPE);
 }
+
+template <typename TI, typename TO>
+shared_ptr<op::Constant> fold_constant_convert_helper1(shared_ptr<op::Constant> constant,
+                                                       const element::Type& output_element_type)
+{
+    auto out_shape = constant->get_shape();
+    vector<TO> out_vec(shape_size(out_shape));
+
+    runtime::reference::convert<TI, TO>(
+        constant->get_vector<TI>().data(), out_vec.data(), shape_size(out_shape));
+
+    return make_shared<op::Constant>(output_element_type, out_shape, out_vec);
+}
+
+template <typename TI>
+shared_ptr<op::Constant> fold_constant_convert_helper0(shared_ptr<op::Constant> constant,
+                                                       const element::Type& output_element_type)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic error "-Wswitch"
+#pragma GCC diagnostic error "-Wswitch-enum"
+    switch (output_element_type.get_type_enum())
+    {
+    case element::Type_t::undefined:
+        NGRAPH_CHECK(false, "Encountered 'undefined' element type in fold_constant_convert");
+        break;
+    case element::Type_t::dynamic:
+        NGRAPH_CHECK(false, "Encountered 'dynamic' element type in fold_constant_convert");
+        break;
+    case element::Type_t::boolean:
+        return fold_constant_convert_helper1<TI, char>(constant, output_element_type);
+    case element::Type_t::bf16:
+        return fold_constant_convert_helper1<TI, bfloat16>(constant, output_element_type);
+    case element::Type_t::f16:
+        return fold_constant_convert_helper1<TI, float16>(constant, output_element_type);
+    case element::Type_t::f32:
+        return fold_constant_convert_helper1<TI, float>(constant, output_element_type);
+    case element::Type_t::f64:
+        return fold_constant_convert_helper1<TI, double>(constant, output_element_type);
+    case element::Type_t::i8:
+        return fold_constant_convert_helper1<TI, int8_t>(constant, output_element_type);
+    case element::Type_t::i16:
+        return fold_constant_convert_helper1<TI, int16_t>(constant, output_element_type);
+    case element::Type_t::i32:
+        return fold_constant_convert_helper1<TI, int32_t>(constant, output_element_type);
+    case element::Type_t::i64:
+        return fold_constant_convert_helper1<TI, int64_t>(constant, output_element_type);
+    case element::Type_t::u8:
+        return fold_constant_convert_helper1<TI, uint8_t>(constant, output_element_type);
+    case element::Type_t::u16:
+        return fold_constant_convert_helper1<TI, uint16_t>(constant, output_element_type);
+    case element::Type_t::u32:
+        return fold_constant_convert_helper1<TI, uint32_t>(constant, output_element_type);
+    case element::Type_t::u64:
+        return fold_constant_convert_helper1<TI, uint64_t>(constant, output_element_type);
+    }
+#pragma GCC diagnostic pop
```

Comment:
GCC 4.8.X has some missed with handling some pragmas. Workaround shown in https://github.com/NervanaSystems/ngraph/blob/master/src/ngraph/op/constant.hpp#L311-L313 is needed.